--- a/Modules/Setup  2015-03-15 07:33:09.093498063 +0000
+++ b/Modules/Setup  2015-03-15 07:33:43.436659171 +0000
@@ -204,10 +204,10 @@
 
 # Socket module helper for SSL support; you must comment out the other
 # socket line above, and possibly edit the SSL variable:
-#SSL=/usr/local/ssl
-#_ssl _ssl.c \
-#      -DUSE_SSL -I$(SSL)/include -I$(SSL)/include/openssl \
-#      -L$(SSL)/lib -lssl -lcrypto
+SSL=/build/openssl-TKTK
+_ssl _ssl.c \
+       -DUSE_SSL -I$(SSL)/include -I$(SSL)/include/openssl \
+       -L$(SSL) -lssl -lcrypto
 
 # The crypt module is now disabled by default because it breaks builds
 # on many systems (where -lcrypt is needed), e.g. Linux (I believe).
